### PR TITLE
Fix image picker conversions

### PR DESCRIPTION
**Pull request by BrtqKr**
### Details

There has been a related PR that was seemingly fixing this issue, but at the same time was the cause of another bug related to HEIC files. It turns out that when using `assetRepresentationMode: 'current'` HEIC files were being given .jpg extension without being converted, so the solution was to convert them manually when detecting the magic bytes with heic signature. It's quite a weird workaround, but since jpg conversion for HEIC is a standard approach, the most reasonable solution was to fix the conversion.  

### Fixed Issues

$ https://github.com/Expensify/App/issues/38478
PROPOSAL:

### Tests

Normally I'd upload both files I've been using for the tests, but github does't accept .heic. The one with the squares from the recordings was .heic 

On iOS

1. Go to any chat
2. Press "+"
3. Add attachment
4. Choose from gallery
5. Select 2 given files (one heic, the other one a standard jpg)
6. Verify that both are being sent and displayed as expected
7. Send a standard video in the same way 
8. Verify that the viso is being sent and displayed correctly
9. Send a video exceeding the file size limit
10. Verify that the picker closes without delay, verify that the error is being displayed instantly

On other platforms 

HEIC files are treated as file attachments, but they should be sent without any issues. Also, verify using the same steps that images of different formats are being uploaded with correct previews in the messages.

- [x] Verify that no errors appear in the JS console

### Offline tests

Everything is related to the picker and the format of the images, so the offline behaviour shouldn't be relevant in this case.

### QA Steps

Same as in standard tests, but please make sure that it's working for various types of media. The last time regression was related to .heic format - this time it's working, but it would be good to verify all available data types of the media picked.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/3dfd767a-960e-4df5-b361-ffe5b55d2b00

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/4b828d1f-53e3-42e5-a0d6-d7f489be1941

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/a99ee520-fee0-4d29-bcb8-13ec7d23fa05

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/e95a92ca-4877-49b8-bc0e-74b20623351c

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/bfa8b61c-f9e5-41dd-af4e-7c32d12bd197

https://github.com/user-attachments/assets/446d40af-dcaa-4452-a6f5-807ad964a8c7

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/8fcf77d1-f69e-424a-bba3-4a72c4ca6b94

</details>

### CODE REVIEW

**Review by hoangzinh — COMMENTED** [review]
Sorry for late review. Overall, your approach looks good to me.

File: src/components/AttachmentPicker/index.native.tsx [hunk 1]
```diff
@@ -158,6 +163,37 @@ function AttachmentPicker({type = CONST.ATTACHMENT_PICKER_TYPE.FILE, children, s
...
+
+                    if (targetAsset?.type?.startsWith('image')) {
+                        const fileContent = await RNFS.read(targetAsset.uri, 12, 0, 'base64');
+                        const hexSignature = Array.from(decode(fileContent))
+                            .map((char) => char.charCodeAt(0).toString(16).padStart(2, '0'))
+                            .slice(0, 32)
+                            .join('')
+                            .toUpperCase();
+
+                        const isHEIC = hexSignature.startsWith(CONST.HEIC_SIGNATURE);
```
[the remark this answers is not in the document]
Comment (BrtqKr) [reply src/components/AttachmentPicker/index.native.tsx]: sure

File: src/libs/fileDownload/FileUtils.ts [hunk 2]
```diff
@@ -255,6 +257,24 @@ function validateImageForCorruption(file: FileObject): Promise<void> {
     });
 }
 
+function verifyFileFormat({fileUri, formatSignature}: {fileUri: string; formatSignature: string}) {
```
[the remark this answers is not in the document]
Comment (BrtqKr) [reply src/libs/fileDownload/FileUtils.ts:260]: we can reuse it, but we would need to pass a different magic bytes for each, for now I'd leave it this way. If we need it in the future we could define a map with signatures for each specific type

https://en.wikipedia.org/wiki/List_of_file_signatures

[the remark this answers is not in the document]
Comment (BrtqKr) [reply src/libs/fileDownload/FileUtils.ts:260]: my miss, sorry for that. It's ready now

**Comment by BrtqKr** [comment]
https://github.com/user-attachments/assets/e732e4b7-00c2-4718-95dd-b5424b4bf5aa

It seems to be working as expected. Would mind sharing the file you've used in this recording?

**Comment by BrtqKr** [comment]
@hoangzinh it wasn't a standard heic. According to the signature of the file you've provided, it was ftypmif1 - a multi-image format related to heic. I've added additional signatures for the related types, which should help in this scenario. Also, I've changed the implementation to use fetch for the purpose of compatibility. I think this should be enough.


https://github.com/user-attachments/assets/23a29bb8-11cd-46cd-8c3f-62efc06eb200

**Review by hoangzinh — APPROVED** [review]
LGTM

File: src/components/AttachmentPicker/index.native.tsx [hunk 3]
```diff
@@ -158,12 +160,42 @@ function AttachmentPicker({type = CONST.ATTACHMENT_PICKER_TYPE.FILE, children, s
...
+                    const targetAssetUri = targetAsset?.uri;
+
+                    if (!targetAssetUri) {
+                        return resolve();
+                    }
+
+                    if (targetAsset?.type?.startsWith('image')) {
+                        FileUtils.verifyFileFormat({fileUri: targetAssetUri, formatSignatures: CONST.HEIC_SIGNATURES})
+                            .then((isHEIC) => {
+                                // react-native-image-picker incorrectly changes file extension without transcoding the HEIC file, so we are doing it manually if we detect HEIC signature
```
Comment (puneetlath) [inline src/components/AttachmentPicker/index.native.tsx:173]: Do we plan to fix this upstream also?

File: src/libs/fileDownload/FileUtils.ts [hunk 4]
```diff
@@ -256,6 +256,23 @@ function validateImageForCorruption(file: FileObject): Promise<{width: number; h
     });
 }
 
+function verifyFileFormat({fileUri, formatSignatures}: {fileUri: string; formatSignatures: readonly string[]}) {
```
Comment (puneetlath) [inline src/libs/fileDownload/FileUtils.ts:259]: Could we add a comment explaining what this function is for. It's not super obvious to me.

**Review by puneetlath — CHANGES_REQUESTED** [review]

File: src/components/AttachmentPicker/index.native.tsx (hunk 3 above)
Comment (BrtqKr) [reply src/components/AttachmentPicker/index.native.tsx:173]: I've sent a feature request to the repository. If they think this approach is ok, I'll send a PR  https://github.com/react-native-image-picker/react-native-image-picker/issues/2317

File: src/libs/fileDownload/FileUtils.ts (hunk 4 above)
Comment (BrtqKr) [reply src/libs/fileDownload/FileUtils.ts:260]: ofc

**Review by puneetlath — APPROVED** [review]

**Pull request merged**